### PR TITLE
sql: support polymorphic functions with decimal elements

### PR DESCRIPTION
**Pull request by sploiselle**
redo of #5145

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5147)

### CODE REVIEW

**Comment by benesch** [comment]
Gah, sorry, I've been trying to review this for a few days now but the problem domain really makes my head hurt and I haven't had much focus time. A couple assorted thoughts:

  1. I think implicit casts between arrays whose subtypes are implicitly castable are a thing in Postgres. A function that takes float[] will happily take an int[] instead, for example, and silently implicitly cast to float[]. I think we do this wrong at the moment, ignoring decimals.
  2. `list_1 || list_2` doesn't quite test implicit casts I think, because polymorphic types don't (is this true?) do implicit casts, but require both sides to have _exactly_ the same type. I think we do this right at the moment, ignoring decimals.

The wrinkle here of course is that our decimal types don't work like Postgres and so we probably need special handling so that polymorphic type resolution does a teeny-tiny-bit of Materialize-specific implicit casting _only_ for adjusting decimal scales. 

tl;dr I think the greatest common decimal thing makes sense in the polymorphic type code, but I'm not sure that it makes sense in the cast code. I think the cast code probably should support (1) (implicit list casts) in _general_, and not just for adjusting decimal scales? If so, then we'd be able to localize the `complex_base_eq` and `find_greatest_common_decimal` methods to the polymorphic type resolution code, which will make me sleep a little easier at night.

**Comment by benesch** [comment]
If that doesn't make sense at all—it's late!—ping me and we can sync over Zoom!

**Comment by sploiselle** [comment]
tl;dr you were almost entirely right, and I was able to localize the worrying functions. The longer response turns more and more into word soup the more I edit it, so my apologies––glad to hop on Zoom to sort this out if there are any remaining questions!

1. You were correct here––I assumed the `CastContext::Assignment` restriction was correct (which it wasn't), so my patch to `plan_hypothetical_cast` carved out an unnecessary niche. The correct fix was just to allow `(List, List)` casts with `CastContext::Implicit`, which you suggested.

    This does let us inline the worrying `ScalarType` function into `resolve_polymorphic_types`, so this should be less objectionable now.

2. `list1 || list2` does test implicit casts between `List`s whose simplest type is `Decimal` (e.g. `decimal list list`) because we coerce every argument used in a polymorphic position to the "resolved type" [using an implicit cast](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/plan/func.rs#L867). In the case of `decimal list`s and the like, the resolved type will be the greatest common decimal. This means we might be asking to implicitly cast between differently scaled decimal lists.

    Note that matching differently scaled `decimal`s is the sole case where we lift the restriction that the polymorphic types have the "exact same type" (from our perspective), and is really all that `complex_base_eq` accomplishes.

File: src/repr/src/scalar.rs [hunk 1]
```diff
@@ -791,6 +791,34 @@ impl<'a> ScalarType {
...
+            }
+            (ScalarType::Map { value_type: l }, ScalarType::Map { value_type: r }) => {
+                let common = l.find_greatest_common_decimal(r)?;
+                Some(ScalarType::Map {
+                    value_type: Box::new(common),
+                })
+            }
+            _ => None,
+        }
+    }
```
Comment (benesch) [inline src/repr/src/scalar.rs:821]: This is dead code now right?

File: src/sql/src/plan/func.rs [hunk 2]
```diff
@@ -391,14 +391,59 @@ impl ParamList {
...
+                }
+                (ScalarType::Map { value_type: l }, ScalarType::Map { value_type: r }) => {
+                    let common = find_greatest_common_decimal(l, r)?;
+                    Some(ScalarType::Map {
+                        value_type: Box::new(common),
+                    })
+                }
+                _ => None,
+            }
+        }
```
Comment (benesch) [inline src/sql/src/plan/func.rs:434]: Can you add a note somewhere in here about why decimals get special treatment? Something like this maybe?

> In PostgreSQL, the `numeric` type does not preserve scale information, so polymorphic type resolution will always treat two `numeric` types as equivalent. To match this behavior in Materialize, we special-case equality here so that we can consider decimals with different scales to be equivalent and resolve the polymorphic constraint to the decimal type with the larger scale.

Def feel free to edit that though!

File: test/sqllogictest/list.slt [hunk 3]
```diff
@@ -1703,34 +1703,66 @@ SELECT (LIST[1] || '{2}')::text
 ----
 {1,2}
 
+# Differently scaled decimals are implcitly castable to one another
```
Comment (benesch) [inline test/sqllogictest/list.slt:1706]: nitty nit: implcitly

**Review by benesch — APPROVED** [review]
🙇🏽🙇🏽🙇🏽 thank you! This looks great to me!

File: src/repr/src/scalar.rs (hunk 1 above)
Comment (sploiselle) [reply src/repr/src/scalar.rs:821]: Derp!

File: src/sql/src/plan/func.rs (hunk 2 above)
Comment (sploiselle) [reply src/sql/src/plan/func.rs:434]: Done
